div.go

```diff
@@ -35,4 +53,87 @@ func udivrem2by1(uh, ul, d, reciprocal uint64) (quot, rem uint64) {
 	}
 
 	return qh, r
-}
\ No newline at end of file
+}
+
+func reciprocal3by2(dh, dl uint64) uint64 {
+	// d[1] is d.hi, d[0] is d.lo
+	var d [2]uint64
+	d[0] = dl
+	d[1] = dh
+	v := reciprocal2by1(d[1])
+	p := d[1] * v
+	p += d[0]
+	if p < d[0] {
+		v--
+		if p >= d[1] {
+			v--
+			p -= d[1]
+		}
+		p -= d[1]
+	}
+
+	th, tl := bits.Mul64(v, d[0])
+
+	p += th
+	if p < th {
+		v--
+		if p >= d[1] {
+			if p > d[1] || tl >= d[0] {
+				v--
+			}
+		}
```

Comment:
Looks odd. This is effectively same as:
```suggestion
		if p > d[1] || tl >= d[0] {
			v--
		}
```

Reply:
Note to myself: these conditions needs special test cases.